File: CppSolution/WordSearch.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution 
{
private:
	bool helper(vector<vector<char> > &board, string word, int index, int starti, int startj, vector<vector<bool> > visit)
	{
		if (word.size() == index) return true;
		if (starti < 0 || starti >= board.size() || startj < 0 || startj >= board[0].size() || visit[starti][startj] || word[index] != board[starti][startj])
			return false;
		visit[starti][startj] = true;
		return helper(board, word, index + 1, starti - 1, startj, visit) ||
		       helper(board, word, index + 1, starti + 1, startj, visit) ||
		       helper(board, word, index + 1, starti, startj - 1, visit) ||
		       helper(board, word, index + 1, starti, startj + 1, visit);
	}
public:
    bool exist(vector<vector<char> > &board, string word) 
    {
    	if (word.size() == 0)
    		return true;
    	int m = board.size();
    	if (m == 0)
    		return false;
    	int n = board[0].size();
    	if (n == 0)
    		return false;
        vector<vector<bool> > visit(m, vector<bool>(n, false));
    	for (int i = 0; i < m; i++)
    	{
    		for (int j = 0; j < n; j++)
    		{
    			if (helper(board, word, 0, i, j, visit))
    				return true;
    		}
    	}
    	return false;
    }
};

int main()
{

	return 0;
}

File: JavaSolution/src/FaceBook/WordSearch.java
package FaceBook;

/**
 * Created by xyunpeng on 1/19/16.
 */
public class WordSearch {
    public boolean exist(char[][] board, String word) {
        if (board == null || board.length == 0) {
            return false;
        }


        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[0].length; j++) {
                if (helper(board, word, 0, i, j)) {
                    return true;
                }
            }
        }

        return false;

    }

    private boolean helper(char[][] board, String word, int len, int x, int y) {
        if (len == word.length()) return true;
        if (x < 0 || y < 0 || x >= board.length || y >= board[0].length) {
            return false;
        }

        if (board[x][y] == word.charAt(len)) {
            char c = board[x][y];
            board[x][y] = '#';                                              //这里用这种方法来避免使用visited数组
            boolean result = helper(board, word, len + 1, x + 1, y) ||
                        helper(board, word, len + 1, x, y + 1) ||
                        helper(board, word, len + 1, x - 1, y) ||
                        helper(board, word, len + 1, x, y - 1);
            board[x][y] = c;
            return result;
        }

        return false;

    }
}


//参考了别人家的代码: https://leetcode.com/discuss/60857/simple-solution